### ISSUE TITLE
Client side dynamodb (dynamodb in the browser)

### ISSUE THREAD

**Issue by realalexhomer** [mhart/react-server-routing-example#3]
Hey,
Fellow NYC developer here! Thanks for making this available. I implemented something kind of similar based on your example, but have been stuck on trying to get dynamodb to work in the client.

The AWS link you provide in the project suggests using cognito, but the documentation for doing it with javascript in the browser seems extremely poor. This is especially true when it comes to dealing with unauthenticated users. 

I did manage to get an unauthenticated cognito token, as well as set the permission on the database, but it seems like a very unnecessary pain in the ass. For one it requires making two trips to AWS (one to get a token, the second to hit the DB), and it seemed quite buggy and poorly conceived from my initial experience. 

So in short I was curious if you ever managed to get dynamodb working from the client with the AWS module, and if so what approach you took? If you could link me to an example or just point me in the right direction that would be really awesome. If you've never done this, thats cool too and thanks for the resource.

**Comment by mhart** [comment]
TBH I think Cognito is the only real option unless you want to use access tokens directly – which of course you can do straight from the browser if you want – you'll just need to be super careful with what permissions you grant them.

The wizard has gotten better in the last year or so in case you haven't looked at it since then: https://console.aws.amazon.com/cognito/create/

And then all you need to do is assign `AWS.config.credentials` as documented here: http://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/CognitoIdentityCredentials.html

This tutorial runs through it more but I imagine you've already seen it? https://blogs.aws.amazon.com/javascript/post/Tx1F7FO6GDAIXD3/Authentication-with-Amazon-Cognito-in-the-Browser

Is there anything in particular that you're running into, or is it just that you're finding it a bit messy?

**Comment by mhart** [comment]
The other option of course is to use API Gateway in front of your DynamoDB calls – that provides you with much more control if you so need: https://aws.amazon.com/api-gateway/

**Comment by realalexhomer** [comment]
I did manage to get it working, but it seems quite messy as you said. I don't want to continue to deal with something like cognito if I'm not authenticating the majority of users.

It seems that dynamodb is more meant for either serving data from a backend or serving data on the front end when you need to authenticate all of your users anyway. Kind of surprising that AWS doesn't have a better solution to make GET requests to dynamodb.

My current thinking is that I'll go back and rebuild what I've done with mongodb instead. I appreciate the help though, I spent a lot of time reading your code and it was very helpful. I'm a front end dev just starting to learn about databases so this has been fun.

If I do figure something out with dynamodb I'll be sure to fill you in as it does seem nice aside from this limitation.

**Comment by mhart** [comment]
Ok cool – I'll close this in the meantime. You're right that DynamoDB was initially created for server-side apps and they opened it up to the browser only later. Still, it's good to have there as an option. Good luck serving from mongodb in the browser! 😉
